aws/resource_aws_s3_bucket_test.go

```diff
@@ -663,6 +661,162 @@ func TestAccAWSS3Bucket_UpdateAcl(t *testing.T) {
 	})
 }
 
+func TestAccAWSS3Bucket_UpdateGrant(t *testing.T) {
+	resourceName := "aws_s3_bucket.bucket"
+	ri := acctest.RandInt()
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSS3BucketDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSS3BucketConfigWithGrants(ri),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSS3BucketExists(resourceName),
+					resource.TestCheckResourceAttr(resourceName, "grant.#", "1"),
+					func(s *terraform.State) error {
+						id := s.RootModule().Resources["data.aws_canonical_user_id.current"].Primary.ID
+						gh := fmt.Sprintf("grant.%v", grantHash(map[string]interface{}{
```

Comment:
👍 I like how you're calculating the hash here. I wish I'd thought of that for some tests I've written…